### DISCUSSION TITLE
Version constraint not respected, how is the order of packageRules?

### DISCUSSION THREAD

**Discussion by ceisele-r** [renovatebot/renovate#11849] [Request Help] [answered]
We are having an issue with one package we want to create a version constraint for.

Our `packageRules` look like this:
```
packageRules: [
    {
      groupName: "definitelyTyped monorepo",
      matchPackagePatterns: ["^@types/"],
    },
    ...
    ...
    {
      matchPackageNames: ["node-fetch", "@​types/node-fetch"],
      allowedVersions: "^2.0.0",
    },
]
```

Now there is an update for `@types/node-fetch` to v3 which should - based on the version constaint in the rule below (the one without `groupName` be ignored).

But still renovate is creating the update for `@types/node-fetch` to v3.

Logged is
```
DEBUG: Dependency @types/node-fetch is part of group definitelyTyped monorepo 
...
DEBUG: npm.updateDependency(): devDependencies.@types/node-fetch = ^3.0.2
```
Besides that, no interesting log messages are logged.

Based on the docs [here](https://docs.renovatebot.com/configuration-options/#packagerules), I understand this the following way:

> Important to know: Renovate will evaluate all packageRules and not stop once it gets a first match. Therefore, you should order your packageRules in order of importance so that later rules can override settings from earlier rules if necessary.

=> I understand: Rules later in the rules array (=lower on my screen) have a higher priority than the ones above. So our rule with the version constraint should override the settings from the "group"-Rule.

Isn't this the correct way to understand this behavior and is the priority probably the other way round (the higher on my screen, the higher the priority of the rule)?

Or could it be that having such a grouping rule _and_ a version constraint rule just does not work as I intend at the moment?

**Comment by rarkins** [comment, marked as answer]
The later/lower rules should take precedence. But if you only have one `allowedVersions` rule then it shouldn't be a factor anyway. Can you reproduce it in a public repo?

**Reply by ceisele-r** [reply]
Ok, thank you for the clarification of the priority @rarkins .

Still, I could reproduce the error in a public repo here:
https://github.com/ceisele-r/renovate-version-constraint-demo1

The repo was "initialized" with `node-fetch` and `@types/node-fetch` both in v2.

Then I executed the renovate action the first time and it created the major update branch for `@types/node-fetch` to v3 (which should not have been created based on the constraint in `renovate.config.js`):
https://github.com/ceisele-r/renovate-version-constraint-demo1/runs/3683672168?check_suite_focus=true

Then I executed the renovate action again so it creates the PR for the created branch:
https://github.com/ceisele-r/renovate-version-constraint-demo1/runs/3683691309?check_suite_focus=true

Debug logs are enabled and here is the created PR:
https://github.com/ceisele-r/renovate-version-constraint-demo1/pull/2

**Reply by rarkins** [reply]
Please put the relevant config in `renovate.json` (i.e. repo config) and not the global config and see if the same thing happens.

**Reply by ceisele-r** [reply]
It unfortunately also happens if I put the `packageRules` settings in the `renovate.json` (https://github.com/ceisele-r/renovate-version-constraint-demo1/commit/2d1f08a4b8f6a4394998011fd018abc4afa87c35):
1. run:
https://github.com/ceisele-r/renovate-version-constraint-demo1/runs/3684301227?check_suite_focus=true
2. run:
https://github.com/ceisele-r/renovate-version-constraint-demo1/runs/3684313622?check_suite_focus=true

PR created:
https://github.com/ceisele-r/renovate-version-constraint-demo1/pull/3

**Reply by rarkins** [reply]
You have a special character (zero width space) in your config:

![image](https://user-images.githubusercontent.com/6311784/134502924-2beed5d7-2840-4bcc-8252-447b1c1e45d1.png)

That's why it matches against `node-fetch` but not `@types/node-fetch`. Remove that space from config and the PR will disappear.

**Reply by ceisele-r** [reply]
Woa...thank you very much.
Seems I've copied this in there from somewhere... :( Sorry for overseeing this...

Definitely a good idea to explicitly watch out for this in the future!

Thanks again!
